# 자동차 연비 데이터셋에서 모델별로 연비에 대한 기술통계량

In [1]:
import pandas as pd
import numpy as np

In [3]:
mycar_data = pd.read_csv('C:/Users/USER/Desktop/강의진행/211208_솔트룩스_데이터분석강의/data/mycars.csv')
print(mycar_data.shape)
mycar_data.head()

(234, 11)


manufacturer model  displacement  year  cylinder automatic driving  mpg  \
0         audi    a4           1.8  1999         4      auto       f   18   
1         audi    a4           1.8  1999         4    manual       f   21   
2         audi    a4           2.0  2008         4    manual       f   20   
3         audi    a4           2.0  2008         4      auto       f   21   
4         audi    a4           2.8  1999         6      auto       f   16   

   highway_mileage fuel    class  
0               29    p  compact  
1               29    p  compact  
2               31    p  compact  
3               30    p  compact  
4               26    p  compact

In [4]:
df = mycar_data[['model','mpg']]
df.groupby('model').describe()

mpg                                                 \
                       count       mean       std   min    25%   50%    75%   
model                                                                         
4runner 4wd              6.0  15.166667  0.752773  14.0  15.00  15.0  15.75   
a4                       7.0  18.857143  1.864454  16.0  18.00  18.0  20.50   
a4 quattro               8.0  17.125000  1.807722  15.0  15.75  17.0  18.25   
a6 quattro               3.0  16.000000  1.000000  15.0  15.50  16.0  16.50   
altima                   6.0  20.666667  1.966384  19.0  19.00  20.0  22.50   
c1500 suburban 2wd       5.0  12.800000  1.303840  11.0  12.00  13.0  14.00   
camry                    7.0  19.857143  1.463850  18.0  18.50  21.0  21.00   
camry solara             7.0  19.857143  1.772811  18.0  18.00  21.0  21.00   
caravan 2wd             11.0  15.818182  1.834022  11.0  15.50  16.0  17.00   
civic                    9.0  24.444444  1.943651  21.0  24.00  24.0  25.00   
corolla                  5.0  25.600000  1.673320  24.0  24.00  26.0  26.00   
corvette                 5.0  15.400000  0.547723  15.0  15.00  15.0  16.00   
dakota pickup 4wd        9.0  12.777778  1.986063   9.0  11.00  14.0  14.00   
durango 4wd              7.0  11.857143  1.573592   9.0  11.00  13.0  13.00   
expedition 2wd           3.0  11.333333  0.577350  11.0  11.00  11.0  11.50   
explorer 4wd             6.0  13.666667  0.816497  13.0  13.00  13.5  14.00   
f150 pickup 4wd          7.0  13.000000  1.000000  11.0  13.00  13.0  13.50   
forester awd             6.0  18.833333  0.983192  18.0  18.00  18.5  19.75   
grand cherokee 4wd       8.0  13.500000  2.507133   9.0  12.50  14.0  15.00   
grand prix               5.0  17.000000  1.000000  16.0  16.00  17.0  18.00   
gti                      5.0  20.000000  2.000000  17.0  19.00  21.0  21.00   
impreza awd              8.0  19.625000  0.744024  19.0  19.00  19.5  20.00   
jetta                    9.0  21.222222  4.867694  16.0  19.00  21.0  21.00   
k1500 tahoe 4wd          4.0  12.500000  1.732051  11.0  11.00  12.5  14.00   
land cruiser wagon 4wd   2.0  12.000000  1.414214  11.0  11.50  12.0  12.50   
malibu                   5.0  18.800000  1.923538  17.0  18.00  18.0  19.00   
maxima                   3.0  18.666667  0.577350  18.0  18.50  19.0  19.00   
mountaineer 4wd          4.0  13.250000  0.500000  13.0  13.00  13.0  13.25   
mustang                  9.0  15.888889  1.452966  14.0  15.00  15.0  17.00   
navigator 2wd            3.0  11.333333  0.577350  11.0  11.00  11.0  11.50   
new beetle               6.0  24.000000  6.511528  19.0  20.00  20.5  27.00   
passat                   7.0  18.571429  1.902379  16.0  17.50  18.0  20.00   
pathfinder 4wd           4.0  13.750000  1.258306  12.0  13.50  14.0  14.25   
ram 1500 pickup 4wd     10.0  11.400000  1.505545   9.0  11.00  11.5  12.75   
range rover              4.0  11.500000  0.577350  11.0  11.00  11.5  12.00   
sonata                   7.0  19.000000  1.414214  18.0  18.00  18.0  20.00   
tiburon                  7.0  18.285714  1.603567  16.0  17.00  19.0  19.50   
toyota tacoma 4wd        7.0  15.571429  0.786796  15.0  15.00  15.0  16.00   

                              
                         max  
model                         
4runner 4wd             16.0  
a4                      21.0  
a4 quattro              20.0  
a6 quattro              17.0  
altima                  23.0  
c1500 suburban 2wd      14.0  
camry                   21.0  
camry solara            22.0  
caravan 2wd             18.0  
civic                   28.0  
corolla                 28.0  
corvette                16.0  
dakota pickup 4wd       15.0  
durango 4wd             13.0  
expedition 2wd          12.0  
explorer 4wd            15.0  
f150 pickup 4wd         14.0  
forester awd            20.0  
grand cherokee 4wd      17.0  
grand prix              18.0  
gti                     22.0  
impreza awd             21.0  
jetta     

In [5]:
# 라이브러리
from scipy import stats
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.stats.proportion import proportions_ztest

__이항분포 연습문제__
- 차량의 도장공정에서 광택도 불량이 40% 가정
- 3대의 차량을 임의로 선택했을 때, 불량대수(0,1,2,3)가 나올 확률

In [7]:
n = 3

for i in range(n+1):
    prop = stats.binom.pmf(k = i, n= n, p = 0.4).round(2)
    print('P(X = {0}): {1}'.format(i, prop))

P(X = 0): 0.22
P(X = 1): 0.43
P(X = 2): 0.29
P(X = 3): 0.06


# 귀무가설 vs 대립가설

__귀무가설__: 설정한 가설의 진실성 확률이 적어 처음부터 버릴 것으로 예상되는 가설 (H0)

__대립가설__: 가설검정에서 진위를 검증하고자하는 가설, 귀무가설이 기각될 때 받아들여지는 가설 (H1)

1. 강아지의 평균수명은 13년이라고 알려져 있다. 몇몇 전문가는 13년이 아니라고 주장한다. 따라서, 강아지의 평균수명이 실제로 어떻게 되는지를 검정하고자 한다. 이때 귀무가설과 대립가설을 설정하시오.

2. 전자제품의 불량률은 3%로 알려져 있다. 그런데 불량률에 대한 항의 민원 전화가 많이 와서, 불량률은 3%보다 더 클 수 있다는 소리가 나오고 있다. 이에 불량률이 실제로 어떻게 되는지 검정하고자 한다

3. 어느 공장에서 생산하는 제품의 분산은 7이라고 알려져 있다. 그런데 품질관리팀의 최근 조사에 의하면, 분산은 7보다 작을 수도 있다는 의견이 나왔다. 따라서, 제품의 분산이 실제로 어떠한지를 알아보기 위하여 검정을 시행하고자 한다.

H0: 강아지의 평균수명은 13년이다  
H1: 강아지의 평균수명은 13년이 아니다

H0 전자제품의 불량률은 3%보다 작거나 같다  
H1: 전자제품의 불량률은 3%보다 크다

H0: 제품의 분산은 7보다 크거나 같다  
H1: 제품의 분산은 7보다 작다

가설검정은 '같다','크다','작다'로 이루어집니다.  

In [11]:
# 데이터 불러오기
df = pd.read_csv('C:/Users/USER/Desktop/강의진행/211208_솔트룩스_데이터분석강의/data/data_norm.csv')

# 정규성 검정
x = stats.shapiro(df)
print('The Shapiro-Wilkis Statistic is %.3f and the p-value is %.3f' %x)

# 1-Sample t-test
t_result = stats.ttest_1samp(df, 76.7)
t, p = t_result.statistic.round(3), t_result.pvalue.round(3)
print( "1-Sample t-test ")
print("t검정통계량 : {}".format(t))
print("p-value : {}".format(p))

The Shapiro-Wilkis Statistic is 0.933 and the p-value is 0.480
1-Sample t-test 
t검정통계량 : [0.864]
p-value : [0.41]


In [13]:
# 데이터 불러오기 
df1 = pd.read_csv('C:/Users/USER/Desktop/강의진행/211208_솔트룩스_데이터분석강의/data/data_sampt1.csv')
df2 = pd.read_csv('C:/Users/USER/Desktop/강의진행/211208_솔트룩스_데이터분석강의/data/data_sampt2.csv')

# 정규성 검정
x=stats.shapiro(df1)
print('Shapiro-Wilkis검정통계량은 %.3f, p-value는 %.3f' %x)


y=stats.shapiro(df2)
print('Shapiro-Wilkis검정통계량은 %.3f, p-value는 %.3f' %y)


# 등분산성 검정
stat, p = stats.bartlett(df1['sample'],df2['sample'])
print('등분산검정결과 p-value는 %.3f' % p)

# 2-Sample t-test
t_result = stats.ttest_ind(df1['sample'], df2['sample'], equal_var = True)
t, p = t_result.statistic.round(3), t_result.pvalue.round(3)

print( "2-Sample t-test ")
print( "t검정통계량: {} " .format(t))
print( "p-value: {} " .format(p))

Shapiro-Wilkis검정통계량은 0.909, p-value는 0.062
Shapiro-Wilkis검정통계량은 0.898, p-value는 0.038
등분산검정결과 p-value는 0.642
2-Sample t-test 
t검정통계량: -3.122 
p-value: 0.003 


[정규성 검정]
- 귀무가설(H0): 정규성이 있다. 
- 대립가설(H1): 정규성이 없다.

[등분산성 검정]
- 귀무가설(H0): 등분산성이 있다.(분산의 차이가 없다) 
- 대립가설(H1): 등분산성이 없다.(분산의 차이가 있다)

__결과 해석__
- 집단 A는 정규성을 만족한다.(유의수준 0.05)
- 집단 B는 정규성을 만족하지 않는다. (유의수준 0.05)
- 집단 A와 B는 등분산성을 만족한다.(분산의 차이가 없다)

__고객 만족도 설문조사 A고객과 B고객 모평균 차이가 있는가?__  
H0: A고객과 B고객은 모평균이 같다.  
H1: A고객과 B고객은 모평균이 다르다.

=> P-value가 0.003으로 귀무가설 기각(대립가설 체택)  
=> A고객과 B고객의 모평균은 차이가 존재한다.

In [17]:
# 데이터 불러오기 
df = pd.read_csv('C:/Users/USER/Desktop/강의진행/211208_솔트룩스_데이터분석강의/data/data_pair.csv')

# paired t-test
t_result = stats.ttest_rel(df['before'], df['after'])

t, p = t_result.statistic.round(3), t_result.pvalue.round(3)

print( "Paired t-test ")
print( " t검정통계량 : {} " .format(t))
print( " p-value : {} " .format(p))


Paired t-test 
 t검정통계량 : -5.324 
 p-value : 0.0 


H0: 영어점수의 전후 차이는 없다.(학원 수강 전 영어 평균 == 학원 수강 후 영어 평균)  
H1: 영어점수의 전후 차이는 있다.(학원 수강 전 영어 평균 != 학원 수강 후 영어 평균)

__p-value = 0.0 < 0.05이므로 유의수준 0.05에서 귀무가설을 기각한다.__


: 학원수강 후에 영어점수의 차이는 있다.

----------------------------------

비율검정1
- H0: 전자제품 A에 대해 품질개선 결과 기존 대비 만족도의 차이가 없다.
- H1: 전자제품 A에 대해 품질개선 결과 기존 대비 만족도의 차이가 있다.
- p-value가 0.05보다 크므로 귀무가설을 기각할 수 없다.
- 결론: 결과적으로 전자제품 A에 대해 품질개선 결과 기존 대비 만족도의 차이가 있다고 할 수 없다.
  
  
  
비율검정2
- H0: 두 공장의 불량률은 차이가 없다.
- H1: 두 공장의 불량률은 차이가 있다.
- p-value가 0.05보다 크므로 귀무가설을 기각할 수 없다.
- 결론: 두 공장의 불량률은 차이가 있다고 할 수 없다.
  
  
  
분산검정
- H0: 고객 만족도의 분산은 차이가 없다.
- H1: 고객 만족도의 분산은 차이가 있다.
- p-value가 0.05보다 크므로 귀무가설을 기각할 수 없다.
- 결론: 고객만족도의 분산은 차이가 없다.


# 상관분석

In [18]:
df = pd.read_csv('C:/Users/USER/Desktop/강의진행/211208_솔트룩스_데이터분석강의/data/data_cor.csv')
df.head()

minutes  units
0        1     23
1        2     29
2        3     49
3        4     64
4        4     74

In [21]:
corr, pval = stats.pearsonr(df['minutes'],df['units'])

In [23]:
print('Correlation Analysis')
print('corr: {0:0.3f}'.format(corr))
print('p:{0:0.3f}'.format(pval))

Correlation Analysis
corr: 0.989
p:0.000


=> 부품 수리시간과 부품 수 간에는 매우 강한 연관성(상관성)이 존재한다.

# 카이스퀘어 검정

H0: 공장별로 전자제품을 생산하는 부하는 차이가 없다.  
H1: 공장별로 전자제품을 생산하는 부하는 차이가 있다.  
    
    
p-value(== 0.038)는 0.05 보다 작으므로 귀무가설을 기각한다  
==> 공장별로 전자제품을 생산하는 부하는 차이가 있다.